marlowe/src/Language/Marlowe/SemanticsTypes.hs

```diff
@@ -651,6 +692,11 @@ instance Eq Action where
     Notify obs1 == Notify obs2 = obs1 == obs2
     _ == _ = False
 
+instance Eq a => Eq (Case a) where
+    {-# INLINABLE (==) #-}
+    Case acl cl == Case acr cr                       = acl == acr && cl == cr
+    MerkleizedCase acl bsl == MerkleizedCase acr bsr = acl == acr && bsl == bsr
+    _ == _                                           = False
```

Comment:
What about situations where a `Case` is conceptually equal to a `MerkleizedCase` if the latter deserialized to the former's contract? Does the rest of the code depend upon the semantics of `Eq (Case a)`? Is there a danger that users of Marlowe might assume that merkleized cases could be equal to the equivalent cases?

Reply:
I am not sure I understand the full extent of your question, but I think they are different to all effects, because they take different kinds of inputs, so they are never interchangeable. The semantics originally depended on `Eq`, but not anymore, we just keep it in case it is useful

Reply:
We may be able to protect the users from seeing `MerkleizedCase` all together, by not including it in extended Marlowe, and keeping it limited to core Marlowe

Reply:
I guess my concern was that a merkleization of a contract and the original normal contract would logically embody the same real-world contract, but they wouldn't be `==`. So I guess we have to accept that equality here is about equality of representation.

Reply:
Yes, in that sense there already are many contracts that are equivalent and we treat them as different, like `If TrueObs Close Close` and `Close`, for example.